### ISSUE TITLE
Custom metrics results returned by 'History' callbacks doesnt work properly

### ISSUE THREAD

**Issue by JVD9kh96** [tensorflow/tensorflow#72366]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.17.0

### Custom code

Yes

### Python version

3.10.13

### Current behavior?

I have defined two custom metrics to track SSIM and PSNR in my denoising autoencoder during the training. While the logs printed during the training seems to work properly, it returns weird results at the end by 'History' callback. All the values for training and validation psnr and ssim equal to the last printed value for validtion psnr and ssim, respectively. Like the older versions, I have defined internal variables by using add_variable method. I have also tried with add_weight. Both showed same behaviour. 

By the way, history callback used to return metrics as numpy arrays, but for the custom metrics that I defined, it returns KerasVariable instances.  Is there any fixes for this? This is a bit annoying that some metrics like loss are returned as numpy arrays while the custom metrics are KerasVariable instances. 

### Standalone code to reproduce the issue

```shell
import tensorflow as tf 
import numpy as np 

class SSIMMetric(tf.keras.metrics.Metric):
    """
    Custom SSIM metric for image denoising.

    This class calculates the average Structural Similarity Index Measure (SSIM)
    across all batches during training or evaluation.

    Args:
      name (str, optional): A name for the metric. Defaults to 'ssim'.
      max_val (float, optional): The dynamic range of the images (usually the maximum pixel value).
          Defaults to 255.0 for images in the 0-255 range.
      **kwargs (optional): Additional keyword arguments for the base Metric class.
    """

    def __init__(self, name='ssim', max_val=255.0, **kwargs):
        super(SSIMMetric, self).__init__(name=name, **kwargs)
        self.max_val = max_val
        self.ssim = self.add_variable(shape=(), name='ssim', initializer='zeros')
        self.counter = self.add_variable(shape=(), name='counter', initializer='zeros')

    def update_state(self, y_true, y_pred, sample_weight=None):
        """
        Updates the internal state of the metric with a batch of data.

        Args:
            y_true (tf.Tensor): The ground truth image tensor.
            y_pred (tf.Tensor): The predicted image tensor.
            sample_weight (tf.Tensor, optional): Sample weights (not used in this implementation).
        """
        ssim = tf.reduce_mean(tf.image.ssim(y_true, y_pred, max_val=self.max_val))
        self.ssim.assign((ssim + self.ssim * self.counter) / (self.counter + 1.))
        self.counter.assign_add(1.)
        
    def result(self):
        """
        Calculates and returns the current average SSIM value.

        Returns:
            tf.Tensor: The average SSIM across all batches.
        """
        return self.ssim 

    def reset_states(self):
        """
        Resets the internal state of the metric to zero.
        """
        self.ssim.assign(0.0)
        self.counter.assign(0.0)


class PSNRMetric(tf.keras.metrics.Metric):
    """
    Custom PSNR metric for image denoising.

    This class calculates the average Peak Signal-to-Noise Ratio (PSNR)
    across all batches during training or evaluation.

    Args:
      name (str, optional): A name for the metric. Defaults to 'psnr'.
      max_val (float, optional): The dynamic range of the images (usually the maximum pixel value).
          Defaults to 255.0 for images in the 0-255 range.
      **kwargs (optional): Additional keyword arguments for the base Metric class.
    """
    def __init__(self, name='psnr', max_val=255.0, **kwargs):
        super(PSNRMetric, self).__init__(name=name, **kwargs)
        self.max_val = max_val
        self.psnr = self.add_variable(shape=(), name='psnr', initializer='zeros')
        self.counter = self.add_variable(shape=(), name='counter', initializer='zeros')
        
    def update_state(self, y_true, y_pred, sample_weight=None):
        """
        Updates the internal state of the metric with a batch of data.

        Args:
            y_true (tf.Tensor): The ground truth image tensor.
            y_pred (tf.Tensor): The predicted image tensor.
            sample_weight (tf.Tensor, optional): Sample weights (not used in this implementation).
        """
        psnr = tf.reduce_mean(tf.image.psnr(y_true, y_pred, max_val=self.max_val))
        self.psnr.assign((psnr + self.counter * self.psnr) / (self.counter + 1.))
        self.counter.assign_add(1.)

    def result(self):
        """
        Calculates and returns the current average PSNR value.

        Returns:
            tf.Tensor: The average PSNR across all batches.
        """
        return self.psnr 

    def reset_states(self):
        """
        Resets the internal state of the metric to zero.
        """
        self.psnr.assign(0.0)
        self.counter.assign(0.0)
        
        

(y_train, _), (y_test, _)   = tf.keras.datasets.cifar10.load_data()

y_train, y_test = y_train/255.0, y_test / 255.0 

x_train = y_train + 0.2 * np.random.random((y_train.shape))
x_test  = y_test  + 0.2 * np.random.random((y_test.shape))

input_layer = tf.keras.layers.Input((32, 32, 3))
x = tf.keras.layers.Conv2D(filters=16, kernel_size=(3, 3), 
                           activation=tf.nn.leaky_relu, padding='same')(input_layer)
x = tf.keras.layers.Conv2D(filters=3, kernel_size=(3, 3), 
                           activation='sigmoid', padding='same')(x)

model = tf.keras.models.Model(input_layer, x)

model.compile(loss='mse', metrics=[PSNRMetric(max_val=1.), SSIMMetric(max_val=1.)], optimizer='Adam')

results = model.fit(x_train, y_train, epochs=5, validation_data=(x_test, y_test))

val_psnr = np.array([np.array(metric_val) for metric_val in results.history['val_psnr']])
psnr     = np.array([np.array(metric_val) for metric_val in results.history['psnr']])
print(f'training psnr:\n{psnr}\nval_psnr:\n{val_psnr}')
```

### Relevant log output

```shell
The logs:

Epoch 1/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - loss: 0.0153 - psnr: 22.2309 - ssim: 0.7574 - val_loss: 0.0017 - val_psnr: 27.8734 - val_ssim: 0.9221
Epoch 2/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.0017 - psnr: 28.0592 - ssim: 0.9232 - val_loss: 0.0016 - val_psnr: 28.2505 - val_ssim: 0.9243
Epoch 3/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - loss: 0.0015 - psnr: 28.3847 - ssim: 0.9248 - val_loss: 0.0015 - val_psnr: 28.4067 - val_ssim: 0.9253
Epoch 4/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - loss: 0.0015 - psnr: 28.4954 - ssim: 0.9255 - val_loss: 0.0015 - val_psnr: 28.5810 - val_ssim: 0.9255
Epoch 5/5
1563/1563 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - loss: 0.0015 - psnr: 28.5493 - ssim: 0.9256 - val_loss: 0.0015 - val_psnr: 28.5256 - val_ssim: 0.9256
training psnr:
[28.525625 28.525625 28.525625 28.525625 28.525625]
val_psnr:
[28.525625 28.525625 28.525625 28.525625 28.525625]
```

**Comment by JVD9kh96** [comment]
instead of using tf.keras.metrics.Metric I also tried tf.keras.Metric for defining the metrics, the results were the same

**Comment by JVD9kh96** [comment]
For those who are looking for a quick fix, you can use legacy keras:
`pip install tf_keras`
`import tensorflow as tf 
import numpy as np 
import os 
os.environ['TF_USE_LEGACY_KERAS'] = 'True'
#... rest of the code 
`

**Comment by tilakrayal** [comment]
@JVD9kh96,
Glad the issue was resolved with the tf-keras. Kindly find the [gist](https://colab.research.google.com/gist/tilakrayal/30aa38cea671499a4eae496c36a391e9/untitled2091.ipynb) of it [here](https://colab.research.google.com/gist/tilakrayal/03384cf3ffd968614056e06829009499/untitled2091.ipynb). And tensorflow v2.17 contains Keras3.0, and this is more related to keras please try to raise the issue in keras-team/keras [repo](https://github.com/keras-team/keras/issues) for the quick resolution. Thank you!
